presto-hive/src/main/java/com/facebook/presto/hive/PrestoS3FileSystem.java

```diff
@@ -546,25 +560,45 @@ private static String keyFromPath(Path path)
 
     private AmazonS3Client createAmazonS3Client(URI uri, Configuration hadoopConfig, ClientConfiguration clientConfig)
     {
-        // first try credentials from URI or static properties
-        try {
-            return new AmazonS3Client(new StaticCredentialsProvider(getAwsCredentials(uri, hadoopConfig)), clientConfig, METRIC_COLLECTOR);
-        }
-        catch (IllegalArgumentException ignored) {
+        AWSCredentialsProvider credentialsProvider = null;
+        String s3ClientKey;
+        String keySuffix = null;
+
+        if (this.s3RoleArn != null) {
+            credentialsProvider = new STSAssumeRoleSessionCredentialsProvider(this.s3RoleArn, Long.toString(randomUUID().getLeastSignificantBits(), AWS_ROLE_SESSION_NAME_MAX_LENGTH));
+            keySuffix = this.s3RoleArn; //different roles should have different s3 clients
         }
+        else {
+            // first try credentials from URI or static properties
+            try {
+                S3Credentials credentials = new S3Credentials();
+                credentials.initialize(uri, hadoopConfig);
+                credentialsProvider = new StaticCredentialsProvider(new BasicAWSCredentials(credentials.getAccessKey(), credentials.getSecretAccessKey()));
+            }
+            catch (IllegalArgumentException ignored) {
+            }
 
-        if (useInstanceCredentials) {
-            return new AmazonS3Client(new InstanceProfileCredentialsProvider(), clientConfig, METRIC_COLLECTOR);
+            if (useInstanceCredentials) {
+                credentialsProvider = new InstanceProfileCredentialsProvider();
```

Comment:
credentialsProvider here overrides the assignment on line 576, so forth it breaks the test com.facebook.presto.hive.TestPrestoS3FileSystem.testStaticCredentials(